050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,821 @@
+
+For example:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: my-pool
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 8Gi
+      cpu: 2
+    limits:
+      memory: 8Gi
+      cpu: 4
+  jvmOptions:
+    -Xms: 2048m
+    -Xmx: 2048m
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: true
+  template:
+    podSet:
+      metadata:
+        labels:
+          key1: label1
+    pod:
+      terminationGracePeriodSeconds: 60
+```
+
+The replicas and storage sections will be required in `KafkaNodePool`.
+The other fields will be optional.
+At the same time, the replicas and storage sections will be made optional in the `Kafka` resource since when using node pools, and they will ignored if set.
+
+When the optional field is not present in `KafkaNodePool` and is present in `Kafka.spec.kafka`, it will default to the configured value from the `Kafka` CR.
+So when `KafkaNodePool.spec.<field>` exists, `KafkaNodePool.spec.<field>` will be ignored.
+This will be applied only on the highest level as demonstrated in the example below:
+
+* If the `KafkaNodePool.spec.jvmOptions` does not exist but `Kafka.spec.kafka.jvmOptions` does exist, the values from `Kafka.spec.kafka.jvmOptions` will be used.
+* When `KafkaNodePool.spec.jvmOptions` exists and has `-Xmx: 1024m` and `Kafka.spec.kafka.jvmOptions` exists as well with `-Xms: 512m`, the operator will completely ignore the value from `Kafka.spec.kafka.jvmOptions` and use only the value from `KafkaNodePool.spec.jvmOptions`.
+* When `KafkaNodePool.spec.template` contains only the field `podSet.metadata.labels`, the operator will completely ignore the corresponding values from `Kafka.spec.kafka.template` (for example `podSet.metadata.annotations`, but also for example `pod.metadata.labels`).
+
+In the future, additional fields might be moved from the `Kafka` resource to the `KafkaNodePool` resource and its `.spec` section according to the needs.
+This proposal tries to minimize the effort and the scope and therefore picks up only some of the fields for the initial implementation.
+Since the existing fields in the `Kafka` CR will be used as the _defaults_, new fields can be added to `KafkaNodePool` resource in the future without any backwards compatibility concerns.
+
+### Resource naming
+
+Each node pool will be represented by its own `StrimziPodSet` and pods running the actual nodes.
+The `StrimziPodSet` will be created automatically by the operator based on the `KafkaNodePool` definition - it will not be created by the users.
+The `StrimziPodSet` will be named `<ClusterName>-<PoolName>` - e.g. `my-cluster-big-nodes`.
+The resulting pods will be named `<ClusterName>-<PoolName>-<ID>` - e.g. `my-cluster-big-nodes-5`.
+The naming of the related resources such as Services, Config Maps, or PVCs will follow the pod names.
```

Comment:
I wonder if we should mention that this puts a bound on how long the combined `<ClusterName>-<PoolName>-<ID>` can be (due to the constraints that Kube has on resource names)?